src/main/js/ephox/imagetools/transformations/ImageResizerCanvas.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * ImageResizerCanvas.js
+ *
+ * Released under LGPL License.
+ * Copyright (c) 1999-2015 Ephox Corp. All rights reserved
+ *
+ * License: http://www.tinymce.com/license
+ * Contributing: http://www.tinymce.com/contributing
+ */
+
+/**
+ * Resizes image/canvas using canvas
+ */
+define("ephox/imagetools/transformations/ImageResizerCanvas", [
+    "ephox/imagetools/util/Promise",
+    "ephox/imagetools/util/Conversions",
+    "ephox/imagetools/util/Canvas",
+    "ephox/imagetools/util/ImageSize"
+], function(Promise, Conversions, Canvas, ImageSize) {
+
+    /**
+     * @method scale
+     * @static
+     * @param image {Image|Canvas}
+     * @param dW {Number} Width that the image should be scaled to
+     * @param dH {Number} Height that the image should be scaled to
+     * @returns {Promise}
+     */
+    function scale(image, dW, dH) {
+        var sW = ImageSize.getWidth(image);
+        var sH = ImageSize.getHeight(image);
+        var wRatio = dW / sW;
+        var hRatio = dH / sH;
+
+        if (wRatio < 0.5 || wRatio > 2) {
+            wRatio = wRatio < 0.5 ? 0.5 : 2;
+        }
+        if (hRatio < 0.5 || hRatio > 2) {
+            hRatio = hRatio < 0.5 ? 0.5 : 2;
+        }
+
+        return _scale(image, wRatio, hRatio).then(function(tCanvas) {
+            var tW = ImageSize.getWidth(tCanvas);
+            var tH = ImageSize.getHeight(tCanvas);
+            if (tW == dW && tH == dH) {
+                return tCanvas;
+            } else {
+                return scale(tCanvas, dW, dH);
```

Comment:
This seems bad. You're passing in an image the first time and a canvas after that, relying on the happy accident that `ImageSize` currently gives you the correct value when you pass it a canvas. I realise `context.drawImage` supports canvas as the source, but this still feels wrong to me.